ip_proofs/src/gipa.rs

```diff
@@ -0,0 +1,556 @@
+use algebra::{bytes::ToBytes, fields::Field, to_bytes};
+use digest::Digest;
+use rand::Rng;
+use std::{marker::PhantomData, ops::MulAssign};
+
+use crate::{mul_helper, Error, InnerProductArgumentError};
+use dh_commitments::DoublyHomomorphicCommitment;
+use inner_products::InnerProduct;
+
+pub struct GIPA<IP, LMC, RMC, IPC, D> {
+    _inner_product: PhantomData<IP>,
+    _left_commitment: PhantomData<LMC>,
+    _right_commitment: PhantomData<RMC>,
+    _inner_product_commitment: PhantomData<IPC>,
+    _digest: PhantomData<D>,
+}
+
+pub struct GIPAProof<IP, LMC, RMC, IPC, D>
+where
+    D: Digest,
+    IP: InnerProduct<
+        LeftMessage = LMC::Message,
+        RightMessage = RMC::Message,
+        Output = IPC::Message,
+    >,
+    LMC: DoublyHomomorphicCommitment,
+    RMC: DoublyHomomorphicCommitment<Scalar = LMC::Scalar>,
+    IPC: DoublyHomomorphicCommitment<Scalar = LMC::Scalar>,
+    RMC::Message: MulAssign<LMC::Scalar>,
+    IPC::Message: MulAssign<LMC::Scalar>,
+    RMC::Key: MulAssign<LMC::Scalar>,
+    IPC::Key: MulAssign<LMC::Scalar>,
+    RMC::Output: MulAssign<LMC::Scalar>,
+    IPC::Output: MulAssign<LMC::Scalar>,
```

Comment:
It might be easier to wrap up things into an `GIPAConfig` trait, maybe something like the following:


```rust
pub trait GIPAConfig {
	type Scalar: ...;

	type LMCKey: ...;
	type RMCKey: ...;
	type IPCKey: ...;

	type LMCMsg: ...;
	type RMCMsg: ...;
	type IPCMsg: ...;

	type LMCOutput: ...;
	type RMCOutput: ...;
	type IPCOutput: ...;

	type IP: InnerProduct<LeftMessage = Self::LMCMsg, RightMessage = Self::RMCMsg, Output = Self::IPCMsg>;

	type LMC: ...;
	type RMC: ...;
	type IPC: ...;
}

Reply:
Agreed with your other comment that having a `GIPAConfig` that implements `Clone` would allow other structs that parameterize with `GIPAConfig` to simply derive `Clone`, which would be easier.

Initializing a `GIPAConfig` would be slightly more annoying due to the 9 extra types (Key, Msg, Output) that need to be specified.

I'm going to keep as is for now, just because of inertia to refactor, but if you have a strong preference, I'm happy to make the change.

---

Separately: In ``zexe``, is there a reason ``Group`` doesn't implement ``Mul<Self>`` or even better ``Mul<&Self>``? Currently am using ``MulAssign`` as a workaround.

Reply:
Do you mean `Mul<Scalar>`? It's because we didn't like the idea of right multiplication by the scalar =P